File: 139-word-break/139-word-break.java
class Solution {
    Set<String> visited = new HashSet<>();
    
    public boolean wordBreak(String s, List<String> wordDict) {
        return helper(s, wordDict);
    }
    
    boolean helper(String s, List<String> wd) {
        if (s.isEmpty()) {
            return true;
        }
        
        if (visited.contains(s)) return false;
        
        visited.add(s);
        
        for(String w : wd) {
            if (s.startsWith(w) && helper(s.substring(w.length()), wd)) {
                return true;
            }
        }
        
        return false;
    }
}

File: 139-word-break/139-word-break.scala
import scala.collection.mutable._

object Solution {
  def wordBreak(s: String, wordDict: List[String]): Boolean = {
    val memo: Map[String, Boolean] = Map()
    helper(s, wordDict, memo)
  }

  def helper(
      s: String,
      dict: List[String],
      memo: Map[String, Boolean]
  ): Boolean = {
    if (s.isEmpty) true
    else if (memo.contains(s)) memo.get(s).get
    else {
      var res = false
      dict.foreach { w =>
        if (s.startsWith(w)) {
          res ||= helper(s.substring(w.length), dict, memo)
        }
      }
      memo += (s -> res)
      res
    }
  }
}